### PR TITLE
Support alternative method of passing Presigned params

### PR DESCRIPTION
**Pull request by joehoyle**
Allow passing `?presign={signature params}` to work around lambda not allowing the pre-soign params directly.

### CODE REVIEW

**Review by kovshenin — APPROVED** [review]

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -13,6 +13,12 @@ jobs:
       - uses: actions/setup-node@v3
         with:
           node-version: 18.x
+      - run: env
```
Comment (rmccue) [inline .github/workflows/test.yml:16]: Was this just for testing?

File: tests/test-private-upload.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * Presign a URL for a given key.
+ * @param key
+ * @returns {Promise<string>}
+ */
+async function getPresignedUrlParams( key: string ) : Promise<Args> {
+	const client = new S3Client(
+		{
+			region: process.env.S3_REGION,
+			bucket: process.env.S3_BUCKET,
+		} as S3ClientConfig
```
Comment (rmccue) [inline tests/test-private-upload.ts:19]: Why is this required? Is there a type error here?

File: tests/test-private-upload.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	);
+	const command = new GetObjectCommand( {
+		Bucket: process.env.S3_BUCKET,
+		Key: key,
+	} );
+
+	/**
+	 * Middleware to remove the x-id query string param form the GetObject call.
+	 */
+	const middleware: MiddlewareType<any, any> = ( next: any, context: any ) => async ( args: any ) => {
```
Comment (rmccue) [inline tests/test-private-upload.ts:29]: Can these be more specifically typed?

File: tests/test-private-upload.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		name: 'tests',
+		relation: 'before',
+		toMiddleware: 'awsAuthMiddleware',
+		override: true,
+	} );
+
+	const presignedUrl = new URL( await getSignedUrl( client, command, {
+		expiresIn: 60,
+	} ) );
+	// console.log( presignedUrl );
```
Comment (rmccue) [inline tests/test-private-upload.ts:45]: Debug code, should be removed.

File: tests/test-private-upload.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	let queryStringParameters: Args = {};
+	presignedUrl.searchParams.forEach( ( value, key ) => {
+		queryStringParameters[ key as keyof Args ] = value;
+	} );
+
+	return queryStringParameters;
```
Comment (rmccue) [inline tests/test-private-upload.ts:51]: Maybe: `Object.fromEntries( presignedUrl.searchParams )` ?

File: tests/test-lambda.ts [hunk 6]
```diff
@@ -58,4 +58,13 @@ global.awslambda = {
 	streamifyResponse( handler: StreamifyHandler ): StreamifyHandler {
 		return handler;
 	},
+	HttpResponseStream: {
+		/**
+		 * @param stream
+		 * @param metadata
```
Comment (rmccue) [inline tests/test-lambda.ts:64]: Missing details

**Review by rmccue — CHANGES_REQUESTED** [review]
Minor things, but the debug/test stuff should be removed before merge.

File: .github/workflows/test.yml (hunk 1 above)
Comment (joehoyle) [reply .github/workflows/test.yml:16]: Yeah my bad.

I'm mobile so can't use the suggest feature it seems

File: tests/test-private-upload.ts (hunk 2 above)
Comment (joehoyle) [reply tests/test-private-upload.ts:19]: Yeah this was copied from the lib.ts. It seems the ts types don't account for this style of passing these times for some reason. Mostly just copied it though

File: tests/test-private-upload.ts (hunk 3 above)
Comment (joehoyle) [reply tests/test-private-upload.ts:29]: They could but it's a test so I'm not that worried

File: tests/test-private-upload.ts (hunk 5 above)
Comment (joehoyle) [reply tests/test-private-upload.ts:51]: Ah that's the function I was looking for I think!
